lib/graphql/query/base_execution/value_resolution.rb

```diff
@@ -38,39 +43,40 @@ def result
         class ListResolution < BaseResolution
           # For each item in the list,
           # Resolve it with the "wrapped" type of this list
-          def result
+          def non_null_result
             wrapped_type = field_type.of_type
+            strategy_class = get_strategy_for_kind(wrapped_type.kind)
             value.map do |item|
-              resolved_type = wrapped_type.resolve_type(item)
-              strategy_class = get_strategy_for_kind(resolved_type.kind)
-              inner_strategy = strategy_class.new(item, resolved_type, target, parent_type, ast_field, query, execution_strategy)
+              inner_strategy = strategy_class.new(item, wrapped_type, target, parent_type, ast_field, query, execution_strategy)
               inner_strategy.result
             end
           end
         end
 
+        class HasPossibleTypeResolution < BaseResolution
+          def non_null_result
+            resolved_type = field_type.resolve_type(value)
+            strategy_class = get_strategy_for_kind(resolved_type.kind)
+            inner_strategy = strategy_class.new(value, resolved_type, target, parent_type, ast_field, query, execution_strategy)
+            inner_strategy.result
+          end
+        end
+
         class ObjectResolution < BaseResolution
           # Resolve the selections on this object
-          def result
+          def non_null_result
             resolver = execution_strategy.selection_resolution.new(value, field_type, ast_field.selections, query, execution_strategy)
             resolver.result
           end
         end
 
-        class EnumResolution < BaseResolution
-          # Get the string name for this enum value
-          def result
-            field_type.coerce_result(value)
-          end
-        end
-
         class NonNullResolution < BaseResolution
           # Get the "wrapped" type and resolve the value according to that type
           def result
+            raise GraphQL::ExecutionError, "Type mismatch resolving '#{ast_field.name}'" if value.nil?
```

Comment:
At the moment `ExecutionError`s are the only ones capable of putting useful information (e.g. a `location` key) into the error data, so I think the exception hierarchy needs a larger cleanup. I'll address that in a separate PR unless @rmosolgo has a different opinion